### DISCUSSION TITLE
Can't quit app on macos

### DISCUSSION THREAD

**Discussion by weareallricky** [pyinstaller/pyinstaller#7738] [Help] [answered]
I just used pyinstaller for the very first time (installed via homebrew on MacOS) to package a windowed app, which uses wxPython. Everything seems to build OK. When I launch the bundled app, it runs as expected, however the only way I can fully terminate the process by force-quitting the app. Closing the app's single window, pressing command-q, or gong to the app's file menu to quit doesn't seem to do anything; the app still appears open and I am unable to open it until I force-quit it.

If I manually run my app (eg unpackaged and from the command line), I am able to quit it with no problem. Any idea what might be going on and how I can resolve this?

**Comment by rokm** [comment]
Can you provide a minimal example that exhibits the problem?

**Reply by weareallricky** [reply]
Yes for sure; I am attaching screenshots and the build log containing debug output from the build.

- [buildlog.txt](https://github.com/pyinstaller/pyinstaller/files/11884461/buildlog.txt) - Debug output from building
- Screenshots from pyinstaller build:
-- ![pyinstaller-build-01](https://github.com/pyinstaller/pyinstaller/assets/37904353/344e700b-ec15-42e0-a0f8-8f06efc1f4e4) Shows app open, with option to quit from the app menu
--![pyinstaller-build-02](https://github.com/pyinstaller/pyinstaller/assets/37904353/dbbad23a-6271-499d-ab5a-bcc311c667f5) Shows screen after quitting the app (this happens from the menu, by clicking the close button on the open app, or pressing command-Q); note the windows are closed but the app menu is still present (but menu contains no items), and the icon (the mustache) is still live in the dock
--![pyinstaller-build-03](https://github.com/pyinstaller/pyinstaller/assets/37904353/a42643d7-a5a9-4f1a-a329-e416a77c5865) Shows app still active in fore quit menu and activity monitor; seems to be the only way to fully quit the app

- Screenshots from executing the original script from the command line:
--![cli-execution-01](https://github.com/pyinstaller/pyinstaller/assets/37904353/656004ac-e38f-4cad-8948-92fe82b57f90) Shows GUI successfully opened via directly executing the original script on the command line
--![cli-execution-02](https://github.com/pyinstaller/pyinstaller/assets/37904353/fae839e9-066c-4228-a099-b1ec05b61527) Shows quit option in the menu
--![cli-execution-03](https://github.com/pyinstaller/pyinstaller/assets/37904353/287ff8f0-9597-462f-8039-3652280946f6) Shows successful exit of the application (from clicking the close button in the GUI, selecting the quit option in the menu, or by pressing command-Q)

**Reply by rokm** [reply]
I meant minimal code example, for reproducing the problem.

**Reply by weareallricky** [reply]
Ah - yes, apologies - attached is the actual code, app is invoked by executing fix_namely_report.py

You'll note that in gui.py, I've tried to explicitly destroy the app on close using wxPython's related event bindings, but to no obvious effect - which has left me believing the problem is not related to wxPython but rather pyinstaller.

[scripts.zip](https://github.com/pyinstaller/pyinstaller/files/11884926/scripts.zip)

And here is the pyinstaller command I've been using to build the app:
`pyinstaller --log-level=DEBUG -w --noconfirm -n "PXT Namely Data Cleaner" --icon="mustache.icns" fix_namely_report.py`

**Reply by weareallricky** [reply]
PS I've also tried isolating the event loop by running the GUI in a separate process - the same problem persists. Updated code at the end of fix_namely_report.py looks like:

```py
def run_gui():
    ndc = NamelyDataCleaner()
    app = gui.NDCApp(False)
    frame = gui.NDCFrame(parent=None, title="PXT Namely Data Cleaner", ndc=ndc)
    frame.Show()
    app.MainLoop()

if __name__ == "__main__":
    multiprocessing.freeze_support()
    p = multiprocessing.Process(target=run_gui)
    p.start()
    p.join()
```

**Reply by rokm** [reply]
Hmmm, yeah, it looks like I can reproduce this with homebrew python 3.11 and homebrew wxpython, but not with python.org python 3.11 and pip-installed wxpython...

**Reply by rokm** [reply, marked as answer]
Ah, yes... `libwx_osx_cocoau_core-3.2.0.2.1.dylib` and its symlink `libwx_osx_cocoau_core-3.2.dylib` are collected as hard copies, leading to problems. Same with `libwx_baseu-3.2.0.2.1.dylib` and `libwx_baseu-3.2.dylib`.

And because the links have a level of indirection in between (`/usr/local/opt/wxwidgets/lib/libwx_baseu_net-3.2.dylib` points to `libwx_baseu_net-3.2.0.dylib`, which points to `libwx_baseu_net-3.2.0.2.1.dylib`) and the middle link is not referenced anywhere (and thus not collected), even #7619 couldn't help here (yet...).

So you'll have to fix this manually:
``` 
cd dist/PXT\ Namely\ Data\ Cleaner.app/Contents/MacOS/
ls -l

rm -f libwx_osx_cocoau_core-3.2.dylib
ln -s libwx_osx_cocoau_core-3.2.0.2.1.dylib libwx_osx_cocoau_core-3.2.dylib

rm -f libwx_baseu-3.2.dylib
ln -s libwx_baseu-3.2.0.2.1.dylib libwx_baseu-3.2.dylib

ls -l
```

Or switch to PyPI version of `wxpython`, which has only the fully-versioned library file without symlinks, so this confusion cannot arise.

**Reply by weareallricky** [reply]
Yep, that fixed it. Thanks so much for your help!
